notary-commons/src/main/kotlin/sidechain/iroha/ReliableIrohaChainListener.kt

```diff
@@ -0,0 +1,107 @@
+package sidechain.iroha
+
+import com.github.kittinunf.result.Result
+import com.rabbitmq.client.ConnectionFactory
+import com.rabbitmq.client.Delivery
+import com.rabbitmq.client.GetResponse
+import config.RMQConfig
+import io.reactivex.Observable
+import io.reactivex.subjects.PublishSubject
+import jp.co.soramitsu.iroha.java.IrohaAPI
+import mu.KLogging
+import sidechain.ChainListener
+import kotlin.test.assertNotNull
+
+/**
+ * Rabbit MQ based implementation of [ChainListener]
+ */
+class ReliableIrohaChainListener(
+    private val irohaAPI: IrohaAPI,
+    private val rmqConfig: RMQConfig,
+    private val irohaQueue: String
+) : ChainListener<Pair<iroha.protocol.BlockOuterClass.Block, () -> Unit>> {
+
+    constructor(
+        irohaHost: String,
+        irohaPort: Int,
+        rmqConfig: RMQConfig,
+        irohaQueue: String
+    ) : this(IrohaAPI(irohaHost, irohaPort), rmqConfig, irohaQueue)
+
+    private val factory = ConnectionFactory()
+
+    private val conn by lazy {
+        factory.host = rmqConfig.host
+        factory.newConnection()
+    }
+
+    private val channel by lazy { conn.createChannel() }
+
+    private var consumerTag: String? = null
+
+    init {
+        channel.exchangeDeclare(rmqConfig.irohaExchange, "fanout", true)
+        channel.queueDeclare(irohaQueue, true, false, false, null)
+        channel.queueBind(irohaQueue, rmqConfig.irohaExchange, "")
+        channel.basicQos(1)
+    }
+
+
+    /**
+     * Returns an observable that emits a new block every time it gets it from Iroha
+     */
+    override fun getBlockObservable(autoAck: Boolean): Result<Observable<Pair<iroha.protocol.BlockOuterClass.Block, () -> Unit>>, Exception> {
+
+        val source = PublishSubject.create<Delivery>()
+        val obs: Observable<Delivery> = source
+        val deliverCallback = { consumerTag: String, delivery: Delivery ->
+            source.onNext(delivery)
+        }
+
+        consumerTag = channel.basicConsume(irohaQueue, autoAck, deliverCallback, { _ -> })
+
+        logger.info { "On subscribe to Iroha chain" }
+        return Result.of {
+            obs.map { delivery ->
+                val block = iroha.protocol.BlockOuterClass.Block.parseFrom(delivery.body)
+                logger.info { "New Iroha block arrived. Height ${block.blockV1.payload.height}" }
+                Pair(block, {
+                    if (!autoAck)
+                        channel.basicAck(delivery.envelope.deliveryTag, false)
+                })
+            }
+        }
+    }
+
+    /**
+     * @return a block as soon as it is committed to iroha
+     */
+    override suspend fun getBlock(autoAck: Boolean): Pair<iroha.protocol.BlockOuterClass.Block, () -> Unit> {
+        var resp: GetResponse?
+        do {
+            resp = channel.basicGet(irohaQueue, autoAck)
```

Comment:
Please add some sleep for ~10ms, i forgot

Reply:
Before or after this line?

Reply:
Doesn't matter :D
I would add it after

Reply:
Ok. I'll do this. But what is the purpose of that?

Reply:
Here we have busy waiting, which can load CPU